File: DD_python/Test/test_Cuts.py
import os
import sys

current_dir = os.path.dirname(os.path.abspath(__file__))
parent_dir = os.path.abspath(os.path.join(current_dir, os.pardir))
sys.path.append(parent_dir)

import unittest
from types import SimpleNamespace
from SourceCode.DD import DD
from SourceCode.GraphAlgorithms.MaxFlow.MaxFlow import MaxFlow
from SourceCode.GraphAlgorithms.MaxFlow.MaxFlowGurobi import MaxFlowGurobi
from SourceCode.DDCutGenerators.FlowCuts import FlowCuts
from SourceCode.DDCutGenerators.JointFlowCuts import JointFlowCuts
from SourceCode.DDCutGenerators.CutStrengthening import CutStrengthening
from SourceCode.DDCutGenerators.TargetCut import TargetCut
from Examples.KnapsackInstance.KnapsackProblem import KnapsackProblem
from Examples.KnapsackInstance.KnapsackGurobiClass import KnapsackGurobi
from Examples.GuiStructure import CutType
from Exceptions.MyExceptions import NoCutStrengthening, MaxFlowNotExecuted, NoCutsFound

class CutTest(unittest.TestCase):
    def setUp(self):
        self.variable_length: int = 4
        self.weights: list[int] = [7, 3, 2, 1]
        self.values: list[int] = [1, 1, 1, 1]
        self.capacity: int = 8
        self.initial_state: list[int] = 0
        self.variables: list[str, list[int]] = [(f'x_{i}', [0, 1]) for i in range(1, self.variable_length + 1)]

        # In-memory stand-in for a KnapsackStructure: exposes the attributes
        # that KnapsackGurobi reads (variables, weights, objective_weights,
        # right_side_of_restrictions) without needing to read a file.
        self.knapsack_parameters = SimpleNamespace(
            variables=self.variables,
            weights=self.weights,
            objective_weights=self.values,
            right_side_of_restrictions=self.capacity,
            initial_state=self.initial_state,
        )

        self.knapsack_instance: 'AbstractProblem' = KnapsackProblem(self.knapsack_parameters)
        self.dd_instance: DD = DD(self.knapsack_instance)
        self.dd_instance.create_decision_diagram(False)
    
    def test_optimize_with_joint_flow_cuts(self):
        gurobiInstance: KnapsackGurobi = KnapsackGurobi(self.knapsack_parameters, continuous=True)
        self.assertIsInstance(gurobiInstance, KnapsackGurobi)

        gurobiInstance.create_model()
        gurobiInstance.optimize_with_cuts([self.dd_instance], cut_type=CutType.JOINT_FLOW)
        self.assertEqual([0.2857142857142857, 1.0, 1.0, 1.0], gurobiInstance.model.getAttr('x'))
        self.assertFalse(gurobiInstance.model.ObjVal.is_integer())
    
    def test_max_flow_instance(self):
        max_flow_instance: MaxFlow = MaxFlow(self.dd_instance)
        self.assertIsInstance(max_flow_instance, MaxFlow)
        
        with self.assertRaises(MaxFlowNotExecuted):
            max_flow_instance.get_max_flow()
        self.assertTrue(max_flow_instance.solve_max_flow([0.2857142857142857, 1.0, 1.0, 1.0]))
        self.assertEqual(0.7142857142857143, max_flow_instance.get_max_flow())
        self.assertTrue(max_flow_instance.solve_max_flow([1, 0, 0, 1]))
        self.assertEqual(1, max_flow_instance.get_max_flow())
        self.assertFalse(max_flow_instance.solve_max_flow([1, 1, 1, 1]))
    
    def test_max_flow_gurobi_instance(self):
        max_flow_gurobi_instance: MaxFlowGurobi = MaxFlowGurobi(self.dd_instance)
        self.assertIsInstance(max_flow_gurobi_instance, MaxFlowGurobi)
        with self.assertRaises(MaxFlowNotExecuted):
            max_flow_gurobi_instance.get_max_flow()
        self.assertTrue(max_flow_gurobi_instance.solve_max_flow([0.2857142857142857, 1.0, 1.0, 1.0]))
        self.assertEqual(0.7142857142857143, max_flow_gurobi_instance.get_max_flow())
        self.assertTrue(max_flow_gurobi_instance.solve_max_flow([1, 0, 0, 1]))
        self.assertEqual(1, max_flow_gurobi_instance.get_max_flow())
        self.assertFalse(max_flow_gurobi_instance.solve_max_flow([1, 1, 1, 1]))
    
    def test_create_flow_cuts_instance(self):
        cuts_instance_dd: FlowCuts = FlowCuts(self.dd_instance)
        self.assertIsInstance(cuts_instance_dd, FlowCuts)

    def test_create_joint_flow_cuts_instance(self):
        cuts_instance_dd: JointFlowCuts = JointFlowCuts(self.dd_instance)
        self.assertIsInstance(cuts_instance_dd, JointFlowCuts)

    def test_create_target_cut_instance(self):
        cuts_instance_dd: TargetCut = TargetCut(self.dd_instance)
        self.assertIsInstance(cuts_instance_dd, TargetCut)

    def test_create_cut_strengthening_instance(self):
        cut_strengthening_instance_dd: CutStrengthening = CutStrengthening(self.dd_instance)
        self.assertIsInstance(cut_strengthening_instance_dd, CutStrengthening)

    def test_get_a_flow_cut(self):
        cuts_instance: FlowCuts = FlowCuts(self.dd_instance)
        gurobiInstance: KnapsackGurobi = KnapsackGurobi(self.knapsack_parameters, continuous=True)
        gurobiInstance.create_model()
        gurobiInstance.optimize_with_cuts([self.dd_instance], cut_type=CutType.JOINT_FLOW)

        with self.assertRaises(NoCutsFound):
            cuts_instance.get_cut()
        self.assertTrue(cuts_instance.generate_cut(gurobiInstance.model.getAttr('x')))
        self.assertEqual(([1, 1, 0, 0], 1), cuts_instance.get_cut())
        self.assertEqual(0.7142857142857143, cuts_instance.get_min_cut())

    def test_get_a_joint_flow_cut(self):
        cuts_instance: JointFlowCuts = JointFlowCuts(self.dd_instance)
        gurobiInstance: KnapsackGurobi = KnapsackGurobi(self.knapsack_parameters, continuous=True)
        gurobiInstance.create_model()
        gurobiInstance.optimize_with_cuts([self.dd_instance], cut_type=CutType.JOINT_FLOW)

        with self.assertRaises(NoCutsFound):
            cuts_instance.get_cut()
        self.assertTrue(cuts_instance.generate_cut(gurobiInstance.model.getAttr('x')))
        self.assertEqual(([1, 1, 0, 0], 1), cuts_instance.get_cut())

    def test_get_a_target_cut(self):
        tc: TargetCut = TargetCut(self.dd_instance)
        with self.assertRaises(NoCutsFound):
            tc.get_cut()
        x_bar = [1.0, 1.0, 1.0, 1.0]
        self.assertTrue(tc.generate_cut(x_bar))
        coeffs, constant = tc.get_cut()
        self.assertAlmostEqual(coeffs[0], 2.5, places=10)
        self.assertAlmostEqual(coeffs[1], 2.5, places=10)
        self.assertAlmostEqual(coeffs[2], 0.0, places=10)
        self.assertAlmostEqual(coeffs[3], 0.0, places=10)
        self.assertAlmostEqual(constant, 2.5, places=10)

    def test_optimize_with_target_cuts(self):
        gurobiInstance: KnapsackGurobi = KnapsackGurobi(self.knapsack_parameters, continuous=True)
        gurobiInstance.create_model()
        gurobiInstance.optimize_with_cuts([self.dd_instance], cut_type=CutType.TARGET)
        self.assertEqual([0.2857142857142857, 1.0, 1.0, 1.0], gurobiInstance.model.getAttr('x'))
        self.assertFalse(gurobiInstance.model.ObjVal.is_integer())

    def test_strengthen_a_cut(self):
        cut_strengthening_instance: CutStrengthening = CutStrengthening(self.dd_instance)
        with self.assertRaises(NoCutStrengthening):
            cut_strengthening_instance.get_cut_strengthening()
        self.assertFalse(cut_strengthening_instance.cut_strengthening([1, 1, 0, 0], 1))
        self.assertTrue(cut_strengthening_instance.cut_strengthening([0.5, 1, 1, 0], 2))
        self.assertEqual(([1, 0, 1, 0], 1), cut_strengthening_instance.get_cut_strengthening())

if __name__ == '__main__':
    unittest.main()

File: DD_c_plusplus/Test/test_Cuts.cpp
#include <gtest/gtest.h>
#include "../../Examples/KnapsackInstance/KnapsackProblem.h"
#include "../../Examples/KnapsackInstance/KnapsackGurobiClass.h"
#include "../../Examples/GuiStructure.h"
#include "../../SourceCode/DD.h"
#include "../../SourceCode/DDCutGenerators/FlowCuts.h"
#include "../../SourceCode/DDCutGenerators/CutStrengthening.h"
#include "../../SourceCode/DDCutGenerators/JointFlowCuts.h"
#include "../../SourceCode/DDCutGenerators/TargetCut.h"
#include "GraphAlgorithms/MaxFlow/MaxFlowGurobi.h"

#include <string>
#include <vector>
#include <memory>

namespace fs = filesystem;
using namespace std;

class CutTest : public ::testing::Test {
protected:
    void SetUp() override {
        variable_length = 4;
        int_weights = {7, 3, 2, 1};
        values = {1, 1, 1, 1};
        capacity = 8;
        variables = {
                make_pair("x_1", vector<int>{0, 1}),
                make_pair("x_2", vector<int>{0, 1}),
                make_pair("x_3", vector<int>{0, 1}),
                make_pair("x_4", vector<int>{0, 1})
        };

        knapsack_params = new KnapsackInstance(new int(0), variables, int_weights, values, capacity);
        knapsack_instance = new KnapsackProblem(*knapsack_params);
        dd_instance = new DD(*knapsack_instance);
        dd_instance->create_decision_diagram(false);

    }

    void TearDown() override {
        delete knapsack_instance;
        delete dd_instance;
        delete knapsack_params;
    }

    int variable_length;
    vector<int> int_weights;
    vector<double> values;
    int capacity;
    vector<pair<string, vector<int>>> variables;
    KnapsackInstance* knapsack_params;
    KnapsackProblem* knapsack_instance;
    DD<int>* dd_instance;
};

TEST_F(CutTest, TestOptimizeWithJointFlowCuts) {
    KnapsackGurobiClass* gurobiInstance = nullptr;
    ASSERT_NO_THROW(gurobiInstance = new KnapsackGurobiClass(*knapsack_params, true));
    ASSERT_TRUE(gurobiInstance != nullptr);

    gurobiInstance->create_model();
    gurobiInstance->optimize_with_cuts(vector<DD<int>*>{dd_instance}, CutType::JointFlow);

    vector<double> expected_solution = {0.2857142857142857, 1.0, 1.0, 1.0};
    for (size_t i = 0; i < gurobiInstance->x.size(); ++i) {
        double value = gurobiInstance->x[i].get(GRB_DoubleAttr_X);
        EXPECT_NEAR(expected_solution[i], value, 1e-6);
    }
    EXPECT_FALSE(gurobiInstance->model->get(GRB_DoubleAttr_ObjVal) == static_cast<int>(gurobiInstance->model->get(GRB_DoubleAttr_ObjVal)));
    delete gurobiInstance;
}

TEST_F(CutTest, TestMaxFlowGurobiInstance) {
    MaxFlowGurobi<int>* maxFlowInstance = nullptr;
    ASSERT_NO_THROW(maxFlowInstance = new MaxFlowGurobi<int>(dd_instance));
    ASSERT_TRUE(maxFlowInstance != nullptr);

    ASSERT_THROW(maxFlowInstance->get_max_flow(), MaxFlowNotExecuted);
    EXPECT_TRUE(maxFlowInstance->solve_max_flow({1.0 , 0, 0, 1.0}));
    EXPECT_TRUE(maxFlowInstance->solve_max_flow({0.2857142857142857, 1.0, 1.0, 1.0}));
    EXPECT_DOUBLE_EQ(0.7142857142857143, maxFlowInstance->get_max_flow());

    EXPECT_FALSE(maxFlowInstance->solve_max_flow({1, 1, 1, 1}));
    delete maxFlowInstance;
}

TEST_F(CutTest, TestMaxFlowInstance) {
    MaxFlow<int>* maxFlowInstance = nullptr;
    ASSERT_NO_THROW(maxFlowInstance = new MaxFlow<int>(dd_instance));
    ASSERT_TRUE(maxFlowInstance != nullptr);

    EXPECT_THROW(maxFlowInstance->get_max_flow(), MaxFlowNotExecuted);
    EXPECT_TRUE(maxFlowInstance->solve_max_flow({1.0 , 0, 0, 1.0}));
    EXPECT_TRUE(maxFlowInstance->solve_max_flow({0.2857142857142857, 1.0, 1.0, 1.0}));
    EXPECT_DOUBLE_EQ(0.7142857142857143, maxFlowInstance->get_max_flow());

    EXPECT_FALSE(maxFlowInstance->solve_max_flow({1, 1, 1, 1}));
    delete maxFlowInstance;
}

TEST_F(CutTest, TestCreateFlowCutsInstance) {
    FlowCuts<int>* cuts_instance = nullptr;
    ASSERT_NO_THROW(cuts_instance = new FlowCuts<int>(dd_instance));
    delete cuts_instance;
}

TEST_F(CutTest, TestCreateJointFlowCutsInstance) {
    JointFlowCuts<int>* cuts_instance = nullptr;
    ASSERT_NO_THROW(cuts_instance = new JointFlowCuts<int>(dd_instance));
    delete cuts_instance;
}

TEST_F(CutTest, TestCreateTargetCutInstance) {
    TargetCut<int>* tc = nullptr;
    ASSERT_NO_THROW(tc = new TargetCut<int>(dd_instance));
    ASSERT_TRUE(tc != nullptr);
    ASSERT_EQ((int)tc->omega.size(), 4);
    delete tc;
}

TEST_F(CutTest, TestGetFlowCuts) {
    FlowCuts<int>* cuts_instance = nullptr;
    KnapsackGurobiClass* gurobiInstance = nullptr;

    cuts_instance = new FlowCuts<int>(dd_instance);
    gurobiInstance = new KnapsackGurobiClass(*knapsack_params, true);
    gurobiInstance->create_model();
    gurobiInstance->optimize_with_cuts(vector<DD<int>*>{dd_instance}, CutType::JointFlow);

    EXPECT_THROW(cuts_instance->get_cut(), NoCutsFound);

    vector<double> x_values = {};
    for (size_t i = 0; i < gurobiInstance->x.size(); ++i) {
        x_values.push_back(gurobiInstance->x[i].get(GRB_DoubleAttr_X));
    }

    EXPECT_TRUE(cuts_instance->generate_cut(x_values));

    auto [coefficients, constant] = cuts_instance->get_cut();
    EXPECT_EQ(coefficients, (vector<double>{1, 1, 0, 0}));
    EXPECT_DOUBLE_EQ(constant, 1.0);

    EXPECT_DOUBLE_EQ(cuts_instance->get_min_cut(), 0.7142857142857143);

    delete cuts_instance;
    delete gurobiInstance;
}

TEST_F(CutTest, TestGetJointFlowCuts) {
    JointFlowCuts<int>* cuts_instance = nullptr;
    KnapsackGurobiClass* gurobiInstance = nullptr;

    cuts_instance = new JointFlowCuts<int>(dd_instance);
    gurobiInstance = new KnapsackGurobiClass(*knapsack_params, true);
    gurobiInstance->create_model();
    gurobiInstance->optimize_with_cuts(vector<DD<int>*>{dd_instance}, CutType::JointFlow);

    EXPECT_THROW(cuts_instance->get_cut(), NoCutsFound);

    vector<double> x_values = {};
    for (size_t i = 0; i < gurobiInstance->x.size(); ++i) {
        x_values.push_back(gurobiInstance->x[i].get(GRB_DoubleAttr_X));
    }

    EXPECT_TRUE(cuts_instance->generate_cut(x_values));

    auto [coefficients, constant] = cuts_instance->get_cut();
    EXPECT_EQ(coefficients, (vector<double>{1, 1, 0, 0}));
    EXPECT_DOUBLE_EQ(constant, 1.0);

    delete cuts_instance;
    delete gurobiInstance;
}

TEST_F(CutTest, TestGetTargetCut) {
    TargetCut<int>* tc = new TargetCut<int>(dd_instance);

    EXPECT_THROW(tc->get_cut(), NoCutsFound);

    vector<double> x_bar = {1.0, 1.0, 1.0, 1.0};
    EXPECT_TRUE(tc->generate_cut(x_bar));

    auto [coefficients, constant] = tc->get_cut();
    ASSERT_EQ((int)coefficients.size(), 4);
    EXPECT_NEAR(coefficients[0], 2.5, 1e-9);
    EXPECT_NEAR(coefficients[1], 2.5, 1e-9);
    EXPECT_NEAR(coefficients[2], 0.0, 1e-9);
    EXPECT_NEAR(coefficients[3], 0.0, 1e-9);
    EXPECT_NEAR(constant, 2.5, 1e-9);

    delete tc;
}

TEST_F(CutTest, TestOptimizeWithTargetCuts) {
    KnapsackGurobiClass* gurobiInstance = new KnapsackGurobiClass(*knapsack_params, true);
    gurobiInstance->create_model();
    gurobiInstance->optimize_with_cuts(vector<DD<int>*>{dd_instance}, CutType::Target);

    vector<double> expected_solution = {0.2857142857142857, 1.0, 1.0, 1.0};
    for (size_t i = 0; i < gurobiInstance->x.size(); ++i) {
        double value = gurobiInstance->x[i].get(GRB_DoubleAttr_X);
        EXPECT_NEAR(expected_solution[i], value, 1e-6);
    }
    EXPECT_FALSE(gurobiInstance->model->get(GRB_DoubleAttr_ObjVal) == static_cast<int>(gurobiInstance->model->get(GRB_DoubleAttr_ObjVal)));
    delete gurobiInstance;
}

TEST_F(CutTest, TestCreateCutStrengtheningInstance) {
    CutStrengthening<int>* cut_strengthening_instance = nullptr;
    ASSERT_NO_THROW(cut_strengthening_instance = new CutStrengthening<int>(dd_instance));
    delete cut_strengthening_instance;
}

TEST_F(CutTest, TestGetCutStrengthened) {
    CutStrengthening<int>* cut_strengthening_instance = nullptr;
    cut_strengthening_instance = new CutStrengthening<int>(dd_instance);

    EXPECT_THROW(cut_strengthening_instance->get_cut_strengthening(), NoCutStrengthening);

    EXPECT_FALSE(cut_strengthening_instance->cut_strengthening({1, 1, 0, 0}, 1));
    EXPECT_TRUE(cut_strengthening_instance->cut_strengthening({0.5, 1, 1, 0}, 2));


    auto strengthened_cut = cut_strengthening_instance->get_cut_strengthening();
    const vector<double>& expected_coefficients = {1, 0, 1, 0};
    EXPECT_EQ(strengthened_cut.first.size(), expected_coefficients.size());
    for (size_t i = 0; i < expected_coefficients.size(); ++i) {
        EXPECT_DOUBLE_EQ(strengthened_cut.first[i], expected_coefficients[i]);
    }

    double expected_constant = 1.0;
    EXPECT_DOUBLE_EQ(strengthened_cut.second, expected_constant);

    delete cut_strengthening_instance;
}
